### PR TITLE
feat/acl: add acl support

### PR DESCRIPTION
**Pull request by jiefenghuang**
1. add interface GetFacl and SetFacl
2. implement ACL support for three engines: TKV, Redis, and SQL.
- sync mode and acl
- add new acl
- inherit from default acl
3. vfs interface support

### CODE REVIEW

File: pkg/acl/acl.go [hunk 1]
```diff
@@ -69,12 +69,12 @@ func (es *Entries) Decode(data []byte) {
 
 // Rule acl rule
 type Rule struct {
-	Owner       uint16
-	Group       uint16
-	Mask        uint16
-	Other       uint16
-	NamedUsers  Entries
-	NamedGroups Entries
+	Owner       uint16  `json:"owner"`
```
Comment (jiefenghuang) [inline pkg/acl/acl.go:72]: for json dump

File: pkg/acl/acl.go [hunk 2]
```diff
@@ -108,17 +108,21 @@ func (r *Rule) Decode(buf []byte) {
 	r.Mask = rb.Get16()
 	r.Other = rb.Get16()
 	uCnt := rb.Get32()
-	r.NamedUsers = make([]Entry, uCnt)
-	for i := 0; i < int(uCnt); i++ {
-		r.NamedUsers[i].Id = rb.Get32()
-		r.NamedUsers[i].Perm = rb.Get16()
+	if uCnt > 0 {
```
Comment (jiefenghuang) [inline pkg/acl/acl.go:111]: If not set, keep it as nil. for json marshal and unmarshal result checking in LoadMeta and DumpMeta

File: pkg/meta/base.go [hunk 3]
```diff
@@ -1160,15 +1163,19 @@ func (m *baseMeta) parseAttr(buf []byte, attr *Attr) {
 	attr.Nlink = rb.Get32()
 	attr.Length = rb.Get64()
 	attr.Rdev = rb.Get32()
-	if rb.Left() >= 8 {
+	if rb.Left() >= 16 {
```
Comment (davies) [inline pkg/meta/base.go:1166]: 8

File: pkg/meta/base.go [hunk 4]
```diff
@@ -1160,15 +1163,19 @@ func (m *baseMeta) parseAttr(buf []byte, attr *Attr) {
...
+	if rb.Left() >= 8 {
+		attr.AccessACLId = rb.Get32()
+		attr.DefaultACLId = rb.Get32()
+	}
 	logger.Tracef("attr: %+v -> %+v", buf, attr)
 }
 
 func (m *baseMeta) marshal(attr *Attr) []byte {
-	w := utils.NewBuffer(36 + 24 + 4 + 8)
+	w := utils.NewBuffer(36 + 24 + 4 + 8 + 8)
```
Comment (davies) [inline pkg/meta/base.go:1178]: save the last 8 bytes if no ACL

File: pkg/meta/base.go [hunk 5]
```diff
@@ -1164,11 +1167,19 @@ func (m *baseMeta) parseAttr(buf []byte, attr *Attr) {
...
+		attr.AccessACLId = rb.Get32()
+		attr.DefaultACLId = rb.Get32()
+	}
 	logger.Tracef("attr: %+v -> %+v", buf, attr)
 }
 
 func (m *baseMeta) marshal(attr *Attr) []byte {
-	w := utils.NewBuffer(36 + 24 + 4 + 8)
+	size := uint32(36 + 24 + 4 + 8)
+	if attr.AccessACLId+attr.DefaultACLId > 0 {
```
Comment (davies) [inline pkg/meta/base.go:1179]: `|` is better

File: pkg/meta/interface.go [hunk 6]
```diff
@@ -161,6 +166,9 @@ type Attr struct {
 	Parent    Ino  // inode of parent; 0 means tracked by parentKey (for hardlinks)
 	Full      bool // the attributes are completed or not
 	KeepCache bool // whether to keep the cached page or not
+
+	AccessACLId  uint32 // access ACL id (identical ACL rules share the same access ACL ID.)
```
Comment (davies) [inline pkg/meta/interface.go:170]: `AccessACL` should be enough

File: pkg/meta/redis.go [hunk 7]
```diff
@@ -605,6 +635,10 @@ func (m *redisMeta) totalInodesKey() string {
 	return m.prefix + totalInodes
 }
 
+func (m *redisMeta) aclKey(id uint32) string {
```
Comment (davies) [inline pkg/meta/redis.go:638]: It's better to have all these in single HashMap

File: pkg/meta/sql.go [hunk 8]
```diff
@@ -4168,3 +4304,161 @@ func (m *dbMeta) doTouchAtime(ctx Context, inode Ino, attr *Attr, now time.Time)
...
+		} else if rule.IsMinimal() && aclType == aclAPI.TypeAccess {
+			// remove acl
+			setAttrACLId(attr, aclType, aclAPI.None)
+			// set mode
+			attr.Mode &= 07000
+			attr.Mode |= ((rule.Owner & 7) << 6) | ((rule.Group & 7) << 3) | (rule.Other & 7)
+		} else {
+			// set acl
+			rule.InheritPerms(attr.Mode)
+			aclId, err := m.insertACL(s, rule)
```
Comment (davies) [inline pkg/meta/sql.go:4396]: after `tryLoadMissACLs` ?

File: pkg/meta/sql.go [hunk 9]
```diff
@@ -4168,3 +4304,161 @@ func (m *dbMeta) doTouchAtime(ctx Context, inode Ino, attr *Attr, now time.Time)
...
+		attr := &Attr{}
+		n := &node{Inode: ino}
+		if ok, err := s.ForUpdate().Get(n); err != nil {
+			return err
+		} else if !ok {
+			return syscall.ENOENT
+		}
+		m.parseAttr(n, attr)
+
+		if ctx.Uid() != 0 && ctx.Uid() != attr.Uid {
```
Comment (davies) [inline pkg/meta/sql.go:4379]: check inmutable

File: pkg/meta/tkv.go [hunk 10]
```diff
@@ -3587,3 +3677,151 @@ func (m *kvMeta) doTouchAtime(ctx Context, inode Ino, attr *Attr, now time.Time)
...
+		}
+		aclId = uint32(newId)
+
+		tx.set(m.aclKey(aclId), rule.Encode())
+		m.aclCache.Put(aclId, rule)
+	}
+	return aclId, nil
+}
+
+func (m *kvMeta) tryLoadMissACLs(tx *kvTxn) error {
```
Comment (davies) [inline pkg/meta/tkv.go:3795]: move this into base, have `doGetACLs(id ...)` here

File: pkg/vfs/vfs.go [hunk 11]
```diff
@@ -927,13 +940,24 @@ func (v *VFS) GetXattr(ctx Context, ino Ino, name string, size uint32) (value []
...
+
+	aclType := GetACLType(name)
+	if aclType != acl.TypeNone {
+		if !v.Conf.Format.EnableACL {
+			err = syscall.ENOTSUP
+			return
+		}
+
+		rule := &acl.Rule{}
+		if errno := v.Meta.GetFacl(ctx, ino, aclType, rule); errno != 0 {
```
Comment (SandyXSD) [inline pkg/vfs/vfs.go:952]:
```suggestion
		if err = v.Meta.GetFacl(ctx, ino, aclType, rule); err != 0 {
```

File: pkg/vfs/vfs.go [hunk 12]
```diff
@@ -1123,3 +1155,102 @@ func InitMetrics(registerer prometheus.Registerer) {
...
+//	N * [ tag:16 perm:16 id:32 ]
+//	tag:
+//	  01 - user
+//	  02 - named user
+//	  04 - group
+//	  08 - named group
+//	  10 - mask
+//	  20 - other
+
+func encodeACL(n *acl.Rule) ([]byte, syscall.Errno) {
```
Comment (SandyXSD) [inline pkg/vfs/vfs.go:1173]: meaningless error code

File: pkg/vfs/vfs.go [hunk 13]
```diff
@@ -927,13 +940,24 @@ func (v *VFS) GetXattr(ctx Context, ino Ino, name string, size uint32) (value []
...
+		}
+
+		rule := &acl.Rule{}
+		if errno := v.Meta.GetFacl(ctx, ino, aclType, rule); errno != 0 {
+			return nil, errno
+		}
+		value, err = encodeACL(rule)
+	} else {
+		err = v.Meta.GetXattr(ctx, ino, name, &value)
+		if size > 0 && len(value) > int(size) {
```
Comment (SandyXSD) [inline pkg/vfs/vfs.go:958]: 'size check' should be outside of the if-else block.

File: pkg/vfs/vfs.go [hunk 14]
```diff
@@ -1123,3 +1155,102 @@ func InitMetrics(registerer prometheus.Registerer) {
...
+//	  20 - other
+
+func encodeACL(n *acl.Rule) ([]byte, syscall.Errno) {
+	length := 4 + 24 + uint32(len(n.NamedUsers)+len(n.NamedGroups))*8
+	if n.Mask != 0xFFFF {
+		length += 8
+	}
+	buff := make([]byte, length)
+	w := utils.NewNativeBuffer(buff)
+	w.Put8(2)  // version
```
Comment (SandyXSD) [inline pkg/vfs/vfs.go:1180]:
```suggestion
	w.Put8(acl.Version)  // version
```

File: pkg/meta/tkv.go [hunk 15]
```diff
@@ -3587,3 +3677,151 @@ func (m *kvMeta) doTouchAtime(ctx Context, inode Ino, attr *Attr, now time.Time)
...
+
+			// set mode
+			if aclType == aclAPI.TypeAccess {
+				attr.Mode &= 07000
+				attr.Mode |= ((rule.Owner & 7) << 6) | ((rule.Mask & 7) << 3) | (rule.Other & 7)
+			}
+		}
+
+		// update attr
+		attr.Ctime = now.Unix()
```
Comment (SandyXSD) [inline pkg/meta/tkv.go:3738]: update only when the acl/mode is actually changed.

File: pkg/meta/tkv.go [hunk 16]
```diff
@@ -3587,3 +3677,151 @@ func (m *kvMeta) doTouchAtime(ctx Context, inode Ino, attr *Attr, now time.Time)
...
+
+		if rule.IsEmpty() {
+			// remove acl
+			setAttrACLId(attr, aclType, aclAPI.None)
+		} else if rule.IsMinimal() && aclType == aclAPI.TypeAccess {
+			// remove acl
+			setAttrACLId(attr, aclType, aclAPI.None)
+			// set mode
+			attr.Mode &= 07000
+			attr.Mode |= ((rule.Owner & 7) << 6) | ((rule.Group & 7) << 3) | (rule.Other & 7)
```
Comment (SandyXSD) [inline pkg/meta/tkv.go:3716]: need `inheritPerms` as well?

File: pkg/meta/tkv.go [hunk 17]
```diff
@@ -249,6 +252,16 @@ func (m *kvMeta) dirQuotaKey(inode Ino) []byte {
 	return m.fmtKey("QD", inode)
 }
 
+func (m *kvMeta) aclKey(id uint32) []byte {
```
Comment (SandyXSD) [inline pkg/meta/tkv.go:255]: add comment for it at around Line 192.

File: pkg/meta/tkv.go [hunk 18]
```diff
@@ -854,13 +878,22 @@ func (m *kvMeta) doLookup(ctx Context, parent Ino, name string, inode *Ino, attr
...
 
 func (m *kvMeta) doGetAttr(ctx Context, inode Ino, attr *Attr) syscall.Errno {
-	a, err := m.get(m.inodeKey(inode))
-	if a != nil {
-		m.parseAttr(a, attr)
-	} else if err == nil {
-		err = syscall.ENOENT
-	}
-	return errno(err)
+	return errno(m.txn(func(tx *kvTxn) error {
```
Comment (SandyXSD) [inline pkg/meta/tkv.go:881]: don't use `m.txn` for read-only operations.

File: pkg/meta/tkv.go [hunk 19]
```diff
@@ -3587,3 +3677,151 @@ func (m *kvMeta) doTouchAtime(ctx Context, inode Ino, attr *Attr, now time.Time)
...
+		return 0
+	}
+
+	if !errors.Is(err, errACLNotInCache) {
+		return errno(err)
+	}
+
+	defer m.timeit("GetFacl", time.Now())
+
+	return errno(m.txn(func(tx *kvTxn) error {
```
Comment (SandyXSD) [inline pkg/meta/tkv.go:3757]: don't use `m.txn` for read-only operations.

File: pkg/meta/tkv.go (hunk 16 above)
Comment (jiefenghuang) [reply pkg/meta/tkv.go:3725]: just set acl back here when remove access acl

File: pkg/meta/interface.go (hunk 6 above)
Comment (jiefenghuang) [reply pkg/meta/interface.go:170]: This is the ACL's ID, not the final value, so I think it should remain unchanged.

File: pkg/meta/sql.go (hunk 9 above)
Comment (jiefenghuang) [reply pkg/meta/sql.go:4367]: done

**Comment by SandyXSD** [comment]
@jiefenghuang `Listxattr` should display ACLs as well?

File: pkg/meta/base.go [hunk 20]
```diff
@@ -2793,6 +2816,78 @@ func (m *baseMeta) CheckSetAttr(ctx Context, inode Ino, set uint16, attr Attr) s
 	if st := m.en.doGetAttr(ctx, inode, &cur); st != 0 {
 		return st
 	}
-	_, st := m.mergeAttr(ctx, inode, set, &cur, &attr, time.Now())
+	_, st := m.mergeAttr(ctx, inode, set, &cur, &attr, time.Now(), nil)
 	return st
 }
+
+const aclCounter = "acl"
```
Comment (SandyXSD) [inline pkg/meta/base.go:2823]: define it together with other counters.

File: pkg/meta/base.go [hunk 21]
```diff
@@ -2793,6 +2816,78 @@ func (m *baseMeta) CheckSetAttr(ctx Context, inode Ino, set uint16, attr Attr) s
...
+	if aclType != aclAPI.TypeAccess && aclType != aclAPI.TypeDefault {
+		return syscall.EINVAL
+	}
+
+	if !ino.IsNormal() {
+		return syscall.EPERM
+	}
+
+	defer func() {
+		m.timeit("SetFacl", time.Now())
```
Comment (SandyXSD) [inline pkg/meta/base.go:2873]: `time.Now()` should be evaluated before `doSetFacl`.

File: pkg/meta/tkv.go [hunk 22]
```diff
@@ -468,6 +490,17 @@ func (m *kvMeta) doInit(format *Format, force bool) error {
 			tx.incrBy(m.counterKey("nextInode"), 2)
 			tx.incrBy(m.counterKey("nextChunk"), 1)
 		}
+
+		// cache all acls
+		acls, err := m.scanValues(m.fmtKey("R"), -1, nil)
```
Comment (SandyXSD) [inline pkg/meta/tkv.go:495]: `doInit` is used to create a new volume. Do this in `NewSession`?

Comment (jiefenghuang) [reply pkg/meta/tkv.go]: move acl cache Initialization (engine.cacheACLs) to NewSession

**Review by SandyXSD — APPROVED** [review]

File: pkg/meta/redis.go [hunk 23]
```diff
@@ -605,6 +638,14 @@ func (m *redisMeta) totalInodesKey() string {
...
+	if key, ok := m.aclKeyCache[id]; ok {
+		return key
+	}
+	m.aclKeyCache[id] = fmt.Sprintf("%sacl%d", m.prefix, id)
```
Comment (davies) [inline pkg/meta/redis.go:645]: this is not thread-safe. I think we don't need this.

Comment (jiefenghuang) [reply pkg/meta/redis.go]: I forgot to handle thread-safety here, and I switched it to sync.Map.

File: pkg/meta/tkv.go [hunk 24]
```diff
@@ -249,6 +257,20 @@ func (m *kvMeta) dirQuotaKey(inode Ino) []byte {
 	return m.fmtKey("QD", inode)
 }
 
+func (m *kvMeta) aclKey(id uint32) []byte {
+	if key, ok := m.aclKeyCache[id]; ok {
+		return key
+	}
+	m.aclKeyCache[id] = m.fmtKey("R", id)
```
[the remark this answers is not in the document]
Comment (jiefenghuang) [reply pkg/meta/tkv.go]: Each get acl request also required the generation of an ACL key, but after testing, the impact on CPU and memory usage was minimal, so it has been removed.

File: pkg/meta/redis.go (hunk 7 above)
Comment (jiefenghuang) [reply pkg/meta/redis.go:639]: Each get acl request also required the generation of an ACL key, but after testing, the impact on CPU and memory usage was minimal, so it has been removed.

Comment (jiefenghuang) [reply pkg/meta/redis.go]: done

**Review by davies — APPROVED** [review]

**Pull request merged**